account/recovery.go

```diff
@@ -123,17 +128,254 @@ func recoverInitalState(ctx context.Context, cfg RecoveryConfig) (
 	return accounts, nil
 }
 
+// recreatePossibleAccounts returns a set of potentially valid accounts
+// by generating a set of recovery keys.
+func recreatePossibleAccounts(ctx context.Context, cfg RecoveryConfig) (
+	map[*Account]struct{}, error) {
+
+	possibleAccounts := make(map[*Account]struct{}, cfg.AccountTarget)
+
+	// Prepare the keys we are going to try. Possibly not all of them will
+	// be used.
+	KeyDescriptors, err := GenerateRecoveryKeys(
+		ctx, cfg.AccountTarget, cfg.Wallet,
+	)
+	if err != nil {
+		return possibleAccounts, fmt.Errorf(
+			"error generating keys: %v", err,
+		)
+	}
+
+	for _, keyDes := range KeyDescriptors {
+		secret, err := cfg.Signer.DeriveSharedKey(
+			ctx, cfg.AuctioneerPubKey, &keyDes.KeyLocator,
+		)
+		if err != nil {
+			return possibleAccounts, fmt.Errorf(
+				"error deriving shared key: %v", err,
+			)
+		}
+
+		acc := &Account{
+			TraderKey:     keyDes,
+			AuctioneerKey: cfg.AuctioneerPubKey,
+			Secret:        secret,
+		}
+		possibleAccounts[acc] = struct{}{}
+	}
+
+	return possibleAccounts, nil
+}
+
+// findAccounts tries to find on-chain footprint for the creation of a set
+// of possible pool accounts.
+func findAccounts(cfg RecoveryConfig,
+	possibleAccounts map[*Account]struct{}) []*Account {
+
+	// The process looks like:
+	//     - Fix the batch counter.
+	//     - Fix a block height.
+	//     - Recreate the script for each of the possible accounts.
+	//         - Look for the script in all of the node txs.
+
+	target := cfg.AccountTarget
+	accounts := make([]*Account, 0, target)
+
+	batchKey := cfg.InitialBatchKey
+	// DecrementKey so the first for iteration starts with the
+	// first batchKey value.
+	batchKey = poolscript.DecrementKey(batchKey)
+
+searchLoop:
+	for batchCounter := 0; batchCounter < 5000; batchCounter++ {
```

Comment:
Should extract this as a constant and perhaps mention that currently there are about 1100 batches (Jan 2022) on mainnet. So we'll remember what this number relates to and when we need to update it.